### PR TITLE
ci: testplan generation cleanup and scope overhaul

### PR DESCRIPTION
**Pull request by nashif**
originally tags were introduced to reduce the number of tests we run, so, if a driver is changed, there was no need to build kernel tests and such, so it was exclusion based. This means that we would end up building loads of irrelevant tests still because tags file does only cover a few areas. 
Most importantly it was abstraction based, so it was not about running tests related to what has changed, instead just reduce the number of tests not guaranteeing coverage at all of what the PR is changing.

This new approach is based on the Maintainer file and connecting areas in that file with tests, if a PR changes one area, tests for this specific area is run. There are a few other steps taken to increase coverage and to deal with special cases.

change some code in some subsys or 2, then...
`./scripts/ci/test_plan.py -c HEAD~1.. or whatever`

for example, changed bindesc subsystem: 

```
i9:zephyr(ci_plans_cleanup_2): ./scripts/ci/test_plan.py -c 6d40dca0323..
INFO: Changed files:
INFO: ///////////////////////////
INFO: subsys/bindesc/CMakeLists.txt
INFO: ///////////////////////////
INFO: -------------------  excludes --------------
INFO: files to be ignored: set()
INFO: not resolved files: ['subsys/bindesc/CMakeLists.txt']
INFO: -------------------  tests --------------
INFO: Remaining files: {'subsys/bindesc/CMakeLists.txt'}
INFO: -------------------  archs --------------
INFO: Remaining files: {'subsys/bindesc/CMakeLists.txt'}
INFO: -------------------  boards --------------
INFO: Remaining files: {'subsys/bindesc/CMakeLists.txt'}
INFO: -------------------  areas --------------
INFO: Remaining files: {'subsys/bindesc/CMakeLists.txt'}
INFO: file: subsys/bindesc/CMakeLists.txt
INFO: area Binary Descriptors changed..
INFO: /home/nashif/zephyrproject/zephyr/scripts/twister -c --test-config custom_config.yaml --level custom --save-tests _test_plan_partial.json --integration
Deleting output directory /home/nashif/zephyrproject/zephyr/twister-out
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.5.0-1517-gbbbec63eca7c
INFO    - Using 'zephyr' toolchain.
INFO    - Selecting default platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/testplan.json
INFO    - Writing JSON report _test_plan_partial.json
INFO: Added 17 suites to plan.
INFO: -------------------  post filters --------------
INFO: No twister needed or partial twister run only...
INFO: Total tests gathered: 17
INFO: Total tests to be run (after removing duplicates): 17
INFO: Total nodes to launch: 1
```

### CODE REVIEW

**Comment by fabiobaltieri** [comment]
Nice, MAINTAINER file approach seems like a great plan.

Do we really need the tags? Maybe the plan could just scan for sample/test yaml files under the area and add those to the plan regardless on the tag.

Was also thinking, maybe it would make sense to distinguish between platform changes and subystem changes and behave differently?

File: scripts/get_maintainer.py [hunk 1]
```diff
@@ -403,12 +405,16 @@ def _print_areas(areas):
...
+\ttests: {}
+\ttags: {}
 \tdescription: {}""".format(area.name,
                             area.status,
                             ", ".join(area.maintainers),
                             ", ".join(area.collaborators),
                             ", ".join(area.inform),
                             ", ".join(area.labels),
+                            ", ".join(area.tests),
+                            ", ".join(area.tags),
```
Comment (fabiobaltieri) [inline scripts/get_maintainer.py:417]: I'm a bit confused by the tags here, is this a transient? You are adding handling, then adding some tags to the maintainer file, then removing the tag and only using tests or am I missing something?

**Comment by fabiobaltieri** [comment]
> That is what the script does, no?

Ok fair enough, sorry not familiar with the original code, need to put more time into understanding this, see my question above about tags+tests in the maintainer file.

File: scripts/get_maintainer.py (hunk 1 above)
Comment (nashif) [reply scripts/get_maintainer.py:417]: yes, started with tags but realized those are very random and inconsistent and do not provide direct relation and decided to look into test identifiers instead which are more accurate.

File: scripts/get_maintainer.py [hunk 2]
```diff
@@ -479,7 +485,7 @@ def ferr(msg):
 
     ok_keys = {"status", "maintainers", "collaborators", "inform", "files",
                "files-exclude", "files-regex", "files-regex-exclude",
-               "labels", "description"}
+               "labels", "description", "tests", "tags"}
```
Comment (fabiobaltieri) [inline scripts/get_maintainer.py:488]: tags in this file can be removed right?

Comment (nashif) [reply scripts/get_maintainer.py:488]: can, but leaving it does not hurt, I might have a usecase for this to improve coverage as we move forward.

Comment (fabiobaltieri) [reply scripts/get_maintainer.py:488]: ack

**Review by fabiobaltieri — DISMISSED** [review]
Having a hard time fining anything to comment about on the code and the history could be cleaned a bit :-), I think the whole file could use some restructuring, but so it did before, I think this is functionally a great improvement worth giving it a shot as is.

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
@nashif Needs rebasing.

**Review by erwango — DISMISSED** [review]
I like the idea.

**Review by MaureenHelm — DISMISSED** [review]

**Comment by MaureenHelm** [comment]
@nashif coming back to this?

**Comment by nashif** [comment]
> @nashif coming back to this?

yes, working on that now actually, it is not easy :(

**Pull request closed without merging**